snap_plugin/v1/plugin.py

```diff
@@ -0,0 +1,216 @@
+        concurrency_count (:obj:`int`): ConcurrencyCount is the max number
+            concurrent calls the plugin may take. If there are 5 tasks using the
+            plugin and concurrency count is 2 there will be 3 plugins running.
+        routing_strategy (:py:class:`RoutingStrategy`): RoutingStrategy will
+            override the routing strategy this plugin requires.  The default
+            routing strategy is least-recently-used.
+        exclusive (:obj:`bool`): Exclusive results in a single instance of the
+            plugin running regardless the number of tasks using the plugin.
+        cache_ttl (:obj:`int`): CacheTTL will override the default cache TTL
+            for the provided plugin and the metrics it exposes (default=500ms).
+        rpc_type (:py:class:`RPCType`)> RPC type
+        rpc_version (:obj:`int`): RPC version
+        unsecure (:obj:`bool`): Unsecure
+    """
+    def __init__(self,
+                 type,
+                 name,
+                 version,
+                 concurrency_count=5,
+                 routing_strategy=RoutingStrategy.lru,
+                 exclusive=False,
+                 cache_ttl=None,
+                 rpc_type=RPCType.grpc,
+                 rpc_version=1,
+                 unsecure=True):
+        self.name = name
+        self.version = version
+        self.type = type
+        self.concurrency_count = concurrency_count
+        self.routing_strategy = routing_strategy
+        self.exclusive = exclusive
+        self.cache_ttl = cache_ttl
+        self.rpc_type = rpc_type
+        self.rpc_version = rpc_version
+        self.unsecure = unsecure
+
+
+class Plugin(object):
+    """Abstract base class for Collector, Processor and Publisher plugins.
+
+    Plugin authoris shoud not inherit directly from this class rather they
+    should inherit from :py:class:`snap_plugin.v1.collector.Collector`,
+    :py:class:`snap_plugin.v1.processor.Processor` or
+    :py:class:`snap_plugin.v1.publisher.Publisher`.
+    """
+    __metaclass__ = ABCMeta
+    def __init__(self):
+        self.meta = None
+        self.proxy = None
+        self.stopped = False
+        self.server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
+        self._port = 0
+
+    def stop_plugin(self):
+        """Stops the plugin"""
+        LOG.debug("stopping plugin")
+        self.server.stop(0)
+        LOG.debug("plugin stopped")
+
+    def start_plugin(self):
+        """Starts the Plugin
+
+        Starting a plugin includes:
+            - Finding an available port
+            - Starting the gRPC server
+            - Printing to STDOUT data (JSON) for handshaking with snapd
+        """
+        LOG.debug("plugin start called..")
+        # find available port for grpc server
+        with closing(socket.socket()) as _socket:
+            _socket.bind(('', 0))
+            self._port = _socket.getsockname()[1]
+            self.server.add_insecure_port('[::]:{!s}'.format(self._port))
```

Comment:
I got it to work by using setsockopt(SOL_SOCKET, SO_REUSEADDR, 1) on _socket. Also needed to switch to using `127.0.0.1` instead of `[::]` since that will break if ipv6 is enabled.